File: python/wooden-puzzle-solver.py
#!/usr/bin/env python3
import copy
import os
import sys
import multiprocessing

SPACES = [
    (x,y,z)
    for x in range(3)
    for y in range(3)
    for z in range(3)
    ]


def instantiate_universe():
    x = [0 for n in range(3)]
    y = [copy.deepcopy(x) for n in range(3)]
    return [copy.deepcopy(y) for n in range(3)]


def nodups(items):
    items = iter(items)
    try:
        last = next(items)
        yield last
        while True:
            item = next(items)
            if item == last:
                continue
            else:
                yield item
                last = item
    except StopIteration:
        return


class Dingus:
    def __init__(self, name, cubes, rotate=True):
        self.name = name
        self.cubes = cubes

        if rotate:
            rotations = (
                rotate_z(rotate_xy(self.cubes, n), m)
                for m in range(6)
                for n in range(4)
            )
            rotations = (sorted(cubes) for cubes in rotations)
            rotations = (
                translate(cubes, (-cubes[0][0], -cubes[0][1], -cubes[0][2]))
                for cubes in rotations
            )
            self.rotations = list(nodups(sorted(rotations)))
            sys.stderr.write("Dingus %s: found %d rotations\n" % (self.name, len(self.rotations),))
        else:
            self.rotations = [sorted(self.cubes)]
            sys.stderr.write("Dingus %s: ignoring rotations\n" % (self.name,))

        placements = [
            translate(rotation, (dx, dy, dz))
            for rotation in self.rotations
            for dx in range(3)
            for dy in range(3)
            for dz in range(3)
            ]
        placements = sorted(
            sorted(cubes)
            for cubes in placements
            if valid(cubes)
            )
        self.placements = list(nodups(placements))

        # A map of placements based on their "lowest" cube:
        self.fillers = {}
        for space in SPACES:
            self.fillers[space] = []
        for placement in placements:
            self.fillers[placement[0]].append(placement)

        sys.stderr.write("          found %d placements\n" % (len(self.placements),))


def main():
    dingusL = Dingus("L", [(0,0,0), (1,0,0), (2,0,0), (2,1,0)])
    dingusT = Dingus("T", [(0,0,0), (1,0,0), (1,1,0), (2,0,0)])
    dingusS = Dingus("S", [(0,0,0), (1,0,0), (1,1,0), (2,1,0)], rotate=False)
    dingusR = Dingus("R", [(0,0,0), (0,1,0), (1,1,0)])
    dingi = [dingusL, dingusL, dingusL, dingusL, dingusT, dingusS, dingusR]

    universe = instantiate_universe()
    count = 0

    count += fill(universe, dingi)
    sys.stderr.write("Found %d solutions\n" % count)


def fill(universe, dingi, to_fill=None):
    """Try to fill universe.

    Try to fill universe given that all spaces not in the `empty` list have
    already been filled."""

    count = 0

    if to_fill is None:
        to_fill = list(reversed(SPACES))
    elif not to_fill:
        if is_canonical(universe):
            print_universe(universe)
            count += 1
        return count

    space = to_fill.pop()
    if universe[space[0]][space[1]][space[2]] != 0:
        # Space is already filled; continue to the next space:
        count += fill(universe, dingi, to_fill)
    else:
        # Find a dingus that can fill space. We only need to look
        # among dingus.fillers[space] because any other positions of a
        # dingus would protrude into the spaces that we have already
        # filled.
        last = None
        for i in range(len(dingi)):
            dingus = dingi.pop(i)
            if dingus != last:
                for placement in dingus.fillers[space]:
                    if placeable(universe, placement):
                        place_cubes(universe, placement, 7 - len(dingi))
                        count += fill(universe, dingi, to_fill)
                        place_cubes(universe, placement, 0)
                last = dingus
            dingi.insert(i, dingus)
    to_fill.append(space)
    return count


def rotate_xy(cubes, rotnum):
    if rotnum == 0:
        return cubes
    if rotnum == 1:
        return tuple((lambda cube: [cube[1], -cube[0], cube[2]])(cube) for cube in cubes)
    if rotnum == 2:
        return tuple((lambda cube: [-cube[0], -cube[1], cube[2]])(cube) for cube in cubes)
    if rotnum == 3:
        return tuple((lambda cube: [-cube[1], cube[0], cube[2]])(cube) for cube in cubes)


def rotate_z(cubes, rotnum):
    if rotnum == 0:
        return cubes
    if rotnum == 1:
        return tuple((lambda cube: [cube[0], cube[2], -cube[1]])(cube) for cube in cubes)
    if rotnum == 2:
        return tuple((lambda cube: [cube[0], -cube[1], -cube[2]])(cube) for cube in cubes)
    if rotnum == 3:
        return tuple((lambda cube: [cube[0], -cube[2], cube[1]])(cube) for cube in cubes)
    if rotnum == 4:
        return tuple((lambda cube: [-cube[2], cube[1], cube[0]])(cube) for cube in cubes)
    if rotnum == 5:
        return tuple((lambda cube: [cube[2], cube[1], -cube[0]])(cube) for cube in cubes)


def valid(cubes):
    for cube in cubes:
        if not all(0 <= coord < 3 for coord in cube):
            return False

    return True


def translate(cubes, origin):
    newcubes = []
    for cube in cubes:
        newcubes.append(tuple(cube[i] + origin[i] for i in range(len(cube))))

    return newcubes


def placeable(universe, cubes):
    for cube in cubes:
        if universe[cube[0]][cube[1]][cube[2]] != 0:
            return False

    return True


def place_cubes(universe, cubes, dingnum):
    for cube in cubes:
        universe[cube[0]][cube[1]][cube[2]] = dingnum


def rotate_coord_xy(x, y, z, n):
    if n == 0:
        return (x, y, z)
    if n == 1:
        return (y, -x, z)
    if n == 2:
        return (-x, -y, z)
    if n == 3:
        return (-y, x, z)


def rotate_coord_z(x, y, z, n):
    if n == 0:
        return (x, y, z)
    if n == 1:
        return (x, z, -y)
    if n == 2:
        return (x, -y, -z)
    if n == 3:
        return (x, -z, y)
    if n == 4:
        return (-z, y, x)
    if n == 5:
        return (z, y, -x)


def rotate_world(universe, xy_rot, z_rot):
    rotated = instantiate_universe()
    for x in range(3):
        for y in range(3):
            for z in range(3):
                cx, cy, cz = x - 1, y - 1, z - 1
                rx, ry, rz = rotate_coord_xy(cx, cy, cz, xy_rot)
                rx, ry, rz = rotate_coord_z(rx, ry, rz, z_rot)
                rotated[rx + 1][ry + 1][rz + 1] = universe[x][y][z]
    return rotated


def universe_key(universe):
    return tuple(
        universe[x][y][z]
        for z in range(3)
        for y in range(3)
        for x in range(3)
    )


def is_canonical(universe):
    canonical_key = universe_key(universe)
    for z_rot in range(6):
        for xy_rot in range(4):
            if z_rot == 0 and xy_rot == 0:
                continue
            rotated = rotate_world(universe, xy_rot, z_rot)
            if universe_key(rotated) < canonical_key:
                return False
    return True


def print_universe(universe):
    for y in range(3):
        for x in range(3):
            if x != 0:
                print('     ', end='')
            for z in range(3):
                if z < 2:
                    print(universe[x][y][z], end=' ')
                else:
                    print(universe[x][y][z], end='')
        print()
    print()


if __name__ == '__main__':
    main()

File: go/wooden-puzzle-solver.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"sort"
)

type Coord int8

type Position struct {
	x, y, z Coord
}

var AllPositions []Position

func init() {
	for x := Coord(0); x < 3; x++ {
		for y := Coord(0); y < 3; y++ {
			for z := Coord(0); z < 3; z++ {
				AllPositions = append(AllPositions, Position{x, y, z})
			}
		}
	}
}

type Cube Position

func CubeLess(c1, c2 Cube) bool {
	if c1.x != c2.x {
		return c1.x < c2.x
	} else if c1.y != c2.y {
		return c1.y < c2.y
	} else {
		return c1.z < c2.z
	}
}

type CubeRotation func(Cube) Cube

func Identity(c Cube) Cube {
	return c
}

func RotateXY(c Cube) Cube {
	return Cube{c.y, -c.x, c.z}
}

func RotateYZ(c Cube) Cube {
	return Cube{c.x, c.z, -c.y}
}

func RotateXZ(c Cube) Cube {
	return Cube{-c.z, c.y, c.x}
}

func Translation(dx, dy, dz Coord) func(Cube) Cube {
	return func(c Cube) Cube {
		return Cube{c.x + dx, c.y + dy, c.z + dz}
	}
}

func Compose(rs ...CubeRotation) CubeRotation {
	return func(c Cube) Cube {
		for i := len(rs) - 1; i >= 0; i-- {
			c = rs[i](c)
		}
		return c
	}
}

func Repeat(r CubeRotation, n int) CubeRotation {
	var rn CubeRotation = Identity
	for i := 0; i < n; i++ {
		rn = Compose(r, rn)
	}
	return rn
}

var XYRotations = []CubeRotation{
	Repeat(RotateXY, 0),
	Repeat(RotateXY, 1),
	Repeat(RotateXY, 2),
	Repeat(RotateXY, 3),
}

var ZRotations = []CubeRotation{
	Repeat(RotateYZ, 0),
	Repeat(RotateYZ, 1),
	Repeat(RotateYZ, 2),
	Repeat(RotateYZ, 3),
	Repeat(RotateXZ, 1),
	Repeat(RotateXZ, 3),
}

var AllRotations []CubeRotation

func init() {
	for _, r1 := range XYRotations {
		for _, r2 := range ZRotations {
			AllRotations = append(AllRotations, Compose(r2, r1))
		}
	}
}

type Dingus struct {
	// A map that groups the placements of this Dingus according
	// to the "lowest" cube in that placement:
	placements map[Position]CubeListSet
}

// Return a new CubeListSet containing only the distinct elements of
// the original.
func (set CubeListSet) Distinct() CubeListSet {
	var distinct CubeListSet

	if len(set) != 0 {
		sort.Sort(set)

		last := set[0]
		distinct = append(distinct, last)

		for _, cubes := range set[1:] {
			if !CubeListsEqual(last, cubes) {
				distinct = append(distinct, cubes)
				last = cubes
			}
		}
	}

	return distinct
}

func NewDingus(w *World, name string, rotate bool, cubes CubeList) Dingus {
	// Note that ordering is invarient under translation, so we
	// can sort the cubes as soon as they are rotated.

	var rotated CubeListSet
	if rotate {
		for _, r := range AllRotations {
			cubes := Apply(cubes, r)
			sort.Sort(cubes)
			c := cubes[0]
			cubes = Apply(cubes, Translation(-c.x, -c.y, -c.z))
			rotated = append(rotated, cubes)
		}
		rotated = rotated.Distinct()
		fmt.Fprintf(os.Stderr, "Dingus %s: found %d rotations\n", name, len(rotated))
	} else {
		sort.Sort(cubes)
		rotated = append(rotated, cubes)
		fmt.Fprintf(os.Stderr, "Dingus %s: ignoring rotations\n", name)
	}

	var all CubeListSet
	for _, cubes := range rotated {
		for _, placed := range w.AllTranslations(cubes) {
			all = append(all, placed)
		}
	}

	all = all.Distinct()

	placements := make(map[Position]CubeListSet)
	var count int
	if len(all) != 0 {
		last := all[0]
		placements[Position(last[0])] = append(placements[Position(last[0])], last)
		count++

		for _, cubes := range all[1:] {
			if !CubeListsEqual(last, cubes) {
				placements[Position(cubes[0])] = append(placements[Position(cubes[0])], cubes)
				count++
				last = cubes
			}
		}
	}

	fmt.Fprintf(os.Stderr, "          found %d placements\n", count)

	return Dingus{placements: placements}
}

type CubeList []Cube

func (c Cube) String() string {
	return fmt.Sprintf("(%d,%d,%d)", c.x, c.y, c.z)
}

func (cubes CubeList) String() string {
	var buf bytes.Buffer
	for i, cube := range cubes {
		if i == 0 {
			fmt.Fprintf(&buf, "%s", cube)
		} else {
			fmt.Fprintf(&buf, ",%s", cube)
		}
	}
	return buf.String()
}

func CubeListLess(cubes1, cubes2 CubeList) bool {
	for n := 0; ; n++ {
		if n == len(cubes1) {
			return n != len(cubes2)
		} else if n == len(cubes2) {
			return false
		} else if cubes1[n] != cubes2[n] {
			return CubeLess(cubes1[n], cubes2[n])
		}
	}
	return false
}

// Allow the cubes in a CubeList to be sorted:

func (cubes CubeList) Len() int {
	return len(cubes)
}

func (cubes CubeList) Less(i, j int) bool {
	return CubeLess(cubes[i], cubes[j])
}

func (cubes CubeList) Swap(i, j int) {
	cubes[i], cubes[j] = cubes[j], cubes[i]
}

// Return true if the two CubeLists' cubes are at the exact same
// locations.
func CubeListsEqual(cubes1, cubes2 CubeList) bool {
	if len(cubes1) != len(cubes2) {
		return false
	}
	for i := 0; i < len(cubes1); i++ {
		if cubes1[i] != cubes2[i] {
			return false
		}
	}
	return true
}

type CubeListSet []CubeList

// Allow the cubeLists in a CubeListSet to be sorted:

func (cubeLists CubeListSet) Len() int {
	return len(cubeLists)
}

func (cubeLists CubeListSet) Less(i, j int) bool {
	return CubeListLess(cubeLists[i], cubeLists[j])
}

func (cubeLists CubeListSet) Swap(i, j int) {
	cubeLists[i], cubeLists[j] = cubeLists[j], cubeLists[i]
}

func Apply(cubes CubeList, r CubeRotation) CubeList {
	var ret CubeList
	for _, c := range cubes {
		ret = append(ret, r(c))
	}
	return ret
}

type World [3][3][3]byte

func (w *World) AllTranslations(cubes CubeList) []CubeList {
	var ret []CubeList

	for dx := Coord(0); dx < 3; dx++ {
		for dy := Coord(0); dy < 3; dy++ {
			for dz := Coord(0); dz < 3; dz++ {
				translated := Apply(cubes, Translation(dx, dy, dz))
				if w.Valid(translated) {
					ret = append(ret, translated)
				}
			}
		}
	}
	return ret
}

func (w *World) ValidCube(c Cube) bool {
	return 0 <= c.x && c.x < 3 &&
		0 <= c.y && c.y < 3 &&
		0 <= c.z && c.z < 3
}

func (w *World) Valid(cubes CubeList) bool {
	for _, c := range cubes {
		if !w.ValidCube(c) {
			return false
		}
	}
	return true
}

// Check whether `cubes` can currently be placed in `w`. `cubes` must be valid.
func (w *World) Fits(cubes CubeList) bool {
	for _, c := range cubes {
		if w[c.x][c.y][c.z] != 0 {
			return false
		}
	}
	return true
}

// Place `cubes` in this world. `cubes` must be valid and must be known to fit
// in this world.
func (w *World) Place(cubes CubeList, value byte) {
	for _, c := range cubes {
		w[c.x][c.y][c.z] = value
	}
}

// Erase `d` from this world. `d` must be valid and should previously
// have been placed in this world.
func (w *World) Unplace(cubes CubeList) {
	for _, c := range cubes {
		w[c.x][c.y][c.z] = 0
	}
}

func (w *World) Print(f *bufio.Writer) {
	for y := 0; y < 3; y++ {
		for x := 0; x < 3; x++ {
			if x != 0 {
				f.WriteString("     ")
			}
			for z := 0; z < 3; z++ {
				c := w[x][y][z]
				if c == 0 {
					f.WriteByte('-')
				} else {
					f.WriteByte(c)
				}
				if z != 2 {
					f.WriteByte(' ')
				}
			}
		}
		f.WriteByte('\n')
	}
	f.WriteByte('\n')
}

func (w *World) Fill(dingi []*Dingus, toFill []Position, c byte, solutions chan<- World) {
	// Find the next position that needs filling:
	var p Position
	for {
		if len(toFill) == 0 {
			solutions <- *w
			return
		}

		p = toFill[0]
		toFill = toFill[1:]
		if w[p.x][p.y][p.z] == 0 {
			break
		}
	}

	// Try placing each distinct dingus at `p`:
	var last *Dingus
	for i, d := range dingi {
		if d == nil || d == last {
			continue
		}
		dingi[i] = nil
		for _, cubes := range d.placements[p] {
			if w.Fits(cubes) {
				w.Place(cubes, c)
				w.Fill(dingi, toFill, c + 1, solutions)
				w.Unplace(cubes)
			}
		}
		dingi[i] = d
		last = d
	}
}

func main() {
	var world World

	var DingusL = NewDingus(&world, "L", true, CubeList{Cube{0, 0, 0}, Cube{1, 0, 0}, Cube{2, 0, 0}, Cube{2, 1, 0}})
	var DingusT = NewDingus(&world, "T", true, CubeList{Cube{0, 0, 0}, Cube{1, 0, 0}, Cube{1, 1, 0}, Cube{2, 0, 0}})
	var DingusS = NewDingus(&world, "S", false, CubeList{Cube{0, 0, 0}, Cube{1, 0, 0}, Cube{1, 1, 0}, Cube{2, 1, 0}})
	var DingusR = NewDingus(&world, "R", true, CubeList{Cube{0, 0, 0}, Cube{0, 1, 0}, Cube{1, 1, 0}})

	var Dingi = []*Dingus{&DingusL, &DingusL, &DingusL, &DingusL, &DingusT, &DingusS, &DingusR}

	solutions := make(chan World)
	go func() {
		world.Fill(Dingi, AllPositions, '1', solutions)
		close(solutions)
	}()

	var count int
	f := bufio.NewWriter(os.Stdout)
	for w := range solutions {
		w.Print(f)
		count++
	}
	f.Flush()
	fmt.Fprintf(os.Stderr, "Found %d solutions\n", count)
}
